src/htsbuf.c

```diff
@@ -353,15 +353,50 @@ htsbuf_append_and_escape_xml(htsbuf_queue_t *hq, const char *s)
 {
   const char *c = s;
   const char *e = s + strlen(s);
-  const char *esc;
+  const char *esc = 0;
   int h;
 
   if(e == s)
     return;
 
-  while(1) {
+  while(c<e) {
     h = *c++;
 
+    if (h & 0x80) {
+      // Start of UTF-8.  But we sometimes have bad UTF-8 (#5909).
+      // So validate and handle bad characters.
+
+      // Number of top bits set indicates the total number of  bytes.
+      const int num_bytes =
+        ((h & 240) == 240) ? 4 :
+        ((h & 224) == 224) ? 3 :
+        ((h & 192) == 192) ? 2 : 0;
+
+      if (!num_bytes) {
+        // Completely invalid sequence, so we replace it with a space.
+        htsbuf_append(hq, s, c - s - 1);
+        htsbuf_append(hq, " ", 1);
+        s=c;
+        continue;
+      } else {
+        // Start of valid UTF-8.
+        if (e - c < num_bytes - 1) {
+          // Invalid sequence - too few characters left in buffer for the sequence.
+          // Append what we already have accumulated and ignore remaining characters.
+          htsbuf_append(hq, s, c - s - 1);
+          break;
+        } else {
+          // We should probably check each character in the range is also valid.
```

Comment:
Your intention is to validate and handle bad characters (according to comment on top). This comment actually says that you are not validating them.

Reply:
There are 2 kinds or validation needed: Validation of UTF and validation of the characters in xml. This patch addresses the first one, so invalid UTF will be filtered. Valid UTF that isn't supposed to be in xml isn't filtered, that's what's left to do. So validation is added but not to the extend that it should be. However, this is already a benefit (as in "better than before") so it can be merged in my opinion.